pkg/kubelet/cm/cpumanager/policy_smtaware.go

```diff
@@ -0,0 +1,85 @@
+
+// smtAwarePolicy is a CPU manager policy that is a further refinement of the existing static policy.
+// This policy allocates CPUs exclusively for a container if all the following
+// conditions are met:
+//
+// - The pod QoS class is Guaranteed.
+// - The CPU request is a positive even integer.
+// This policy ensures that containers will be admitted by the kubelet only if the CPU request is even integer.
+// In case CPU request is not even, the pod will not be admitted and will fail with `SMTAlignmentError`.
+// The rational behind this is to ensure that in case of SMT-enabled systems, a guaranteed container never
+// shares a physical core with other containers.
+type smtAwarePolicy struct {
+	*staticPolicy
+}
+
+// Ensure smtAwarePolicy implements Policy interface
+var _ Policy = &smtAwarePolicy{}
+
+// NewsmtAwarePolicy returns a CPU manager policy that does not change CPU
+// assignments for exclusively pinned guaranteed containers after the main
+// container process starts.
+func NewSMTAwarePolicy(topology *topology.CPUTopology, numReservedCPUs int, reservedCPUs cpuset.CPUSet, affinity topologymanager.Store) (Policy, error) {
+	pol, err := NewStaticPolicy(topology, numReservedCPUs, reservedCPUs, affinity)
+	if err != nil {
+		return &smtAwarePolicy{}, err
+	}
+	p, ok := pol.(*staticPolicy)
+	if !ok {
+		klog.ErrorS(err, "unexpected policy type, initialization of smtaware policy requires a static policy")
+		return &smtAwarePolicy{}, err
+	}
+	return &smtAwarePolicy{p}, nil
+}
+
+func (p *smtAwarePolicy) Name() string {
+	return string(PolicySMTAware)
+}
+
+func (p *smtAwarePolicy) Admit(pod *v1.Pod) lifecycle.PodAdmitResult {
+	cpusPerCore := p.topology.CPUsPerCore()
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+		if requestedCPUs := p.guaranteedCPUs(pod, &container); requestedCPUs > 0 && (requestedCPUs%cpusPerCore) != 0 {
+			// Since CPU Manager has been enabled with `smtaware` policy, it means a guaranteed pod can only be admitted
+			// if the CPU requested is a multiple of the number of virtual cpus per physical cores.
+			// In case CPU request is not a multiple of the number of virtual cpus per physical cores the Pod will be put
+			// in Failed state, with SMTAlignmentError as reason. Since the allocation happens in terms of physical cores
+			// and the scheduler is responsible for ensuring that the workload goes to a node that has enough CPUs,
+			// the pod would be placed on a node where there are enough physical cores available to be allocated.
+			// Just like the behaviour in case of static policy, takeByTopology will try to first allocate CPUs from the same socket
+			// and only in case the request cannot be sattisfied on a single socket, CPU allocation is done for a workload to occupy all
+			// CPUs on a physical core. Allocation of individual threads would never have to occur.
+			return smtAlignmentError(requestedCPUs, cpusPerCore)
+		}
+	}
```

Comment:
I think this logic is better served as an error in the call to `Allocate()` (which then triggers an admission error since no allocation can be made).

Reply:
Our initial approach was indeed to perform this check in the call to `Allocate` but pivoted to do this at admission stage because it implicitly meant that we would be admitting pods that could result in `SMTAlignment` error eventually. Having said that, we have no issues moving back to performing the check in Allocate but wanted to explain the reason we thought performing in Admit was better.

Reply:
It seems the only advantage having an explicit `Admit()` handler gives us, is the ability to more precisely set the “Reason”, which I think could be fixed by updating the way Allocate() propagates its errors back.

I’d also be fine *only* having an `Admit()` handler and not have failures to `Allocate()` trigger an admission error (but then I’d want that to be consistent across all of the components that currently rely on `Allocate()` for this). 

It just seems weird to have an explicit `Admit()` handler for `SMTAlignment`, and then leave all other admission errors to the `Allocate()` call.

Reply:
Got it! Thanks for the clarification.